compiler/noirc_evaluator/src/ssa/opt/constant_folding/mod.rs

```diff
@@ -111,18 +116,45 @@ impl Function {
     pub(crate) fn constant_fold(
         &mut self,
         use_constraint_info: bool,
+        max_iter: usize,
         interpreter: &mut Option<Interpreter<Empty>>,
     ) {
-        let mut context = Context::new(use_constraint_info);
         let mut dom = DominatorTree::with_function(self);
+        let mut context = Context::new(use_constraint_info);
         context.block_queue.push_back(self.entry_block());
 
-        while let Some(block) = context.block_queue.pop_front() {
-            context.fold_constants_in_block(&mut self.dfg, &mut dom, block, interpreter);
-        }
+        for _ in 0..max_iter {
+            while let Some(block) = context.block_queue.pop_front() {
+                context.fold_constants_in_block(&mut self.dfg, &mut dom, block, interpreter);
+                context.block_queue.extend(self.dfg[block].successors());
+            }
+
+            #[cfg(debug_assertions)]
+            constant_folding_post_check(&context, &self.dfg);
 
-        #[cfg(debug_assertions)]
-        constant_folding_post_check(&context, &self.dfg);
+            // In order to deduplicate and hoist new instructions, we need rebuild the cache starting from the
+            // blocks we hoisted into, revisiting all descendants.
+            let blocks_to_revisit = std::mem::take(&mut context.blocks_to_revisit);
+
+            // Find blocks which are not dominated by anything other than themselves; these are our independent starting points.
+            // Alternatively we could reduce all blocks to their common dominator.
```

Comment:
Did you explore using the common dominator? Did it cause us to regress significantly in the number of restarts? I feel in some scenarios we could in fact have worse performance with multiple restart points if they share exit paths.